// solc-typed-ast/dist/ast/implementation/expression/function_call_options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionCallOptions = void 0;
const expression_1 = require("./expression");
class FunctionCallOptions extends expression_1.Expression {
    constructor(id, src, typeString, expression, options, raw) {
        super(id, src, typeString, raw);
        this.vExpression = expression;
        this.vOptionsMap = options;
        this.acceptChildren();
    }
    get children() {
        return this.pickNodes(this.vExpression, this.vOptions);
    }
    /**
     * Specified option names, respecting order. For following example:
     * ```
     * foreign.buy{gas: 2000, value: 1 gwei}(1000);
     * ```
     * Will contain `["gas", "value"]`.
     */
    get names() {
        return this.vOptionsMap.keys();
    }
    /**
     * Specified option values, respecting order. For following example:
     * ```
     * foreign.buy{gas: 2000, value: 1 gwei}(1000);
     * ```
     * Will contain `[2000, 1 gwei]` literal expressions.
     */
    get vOptions() {
        return this.vOptionsMap.values();
    }
}
exports.FunctionCallOptions = FunctionCallOptions;
