// solc-typed-ast/dist/ast/implementation/expression/member_access.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemberAccess = void 0;
const expression_1 = require("./expression");
class MemberAccess extends expression_1.Expression {
    constructor(id, src, typeString, expression, memberName, referencedDeclaration, raw) {
        super(id, src, typeString, raw);
        this.vExpression = expression;
        this.memberName = memberName;
        this.referencedDeclaration = referencedDeclaration;
        this.acceptChildren();
    }
    get children() {
        return this.pickNodes(this.vExpression);
    }
    /**
     * Reference to the declaration
     */
    get vReferencedDeclaration() {
        return this.requiredContext.locate(this.referencedDeclaration);
    }
    set vReferencedDeclaration(value) {
        if (value === undefined) {
            this.referencedDeclaration = -1;
        }
        else {
            if (!this.requiredContext.contains(value)) {
                throw new Error(`Node ${value.type}#${value.id} not belongs to a current context`);
            }
            this.referencedDeclaration = value.id;
        }
    }
}
exports.MemberAccess = MemberAccess;
