// solc-typed-ast/dist/misc/location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Location = void 0;
const source_unit_1 = require("../ast/implementation/meta/source_unit");
const SrcLocation = require("src-location");
class Location {
    constructor(fileName, src) {
        this.fileName = fileName;
        this.src = src;
    }
    static createForNode(node) {
        const root = node.root;
        if (root instanceof source_unit_1.SourceUnit) {
            return new Location(root.sourceEntryKey, node.src);
        }
        throw new Error(`Unable to create location for node (${node.type} #${node.id}) the root of which is not a source unit`);
    }
    getCoordinates(content) {
        if (content === undefined) {
            return {
                start: {
                    line: 0,
                    column: 0
                },
                end: {
                    line: 0,
                    column: 0
                }
            };
        }
        const coords = this.src.split(":");
        const index = parseInt(coords[0], 10);
        const length = parseInt(coords[1], 10);
        const start = SrcLocation.indexToLocation(content, index, true);
        const end = SrcLocation.indexToLocation(content, index + length, true);
        return { start, end };
    }
}
exports.Location = Location;
