// solc-typed-ast/dist/types/reserved.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = void 0;
const constants_1 = require("../ast/constants");
const ast_1 = require("./ast");
// Helper with some singleton types to avoid unnecessary allocations
exports.types = {
    bool: new ast_1.BoolType(),
    uint256: new ast_1.IntType(256, false),
    uint160: new ast_1.IntType(160, false),
    uint8: new ast_1.IntType(8, false),
    byte: new ast_1.FixedBytesType(1),
    bytes4: new ast_1.FixedBytesType(4),
    bytes20: new ast_1.FixedBytesType(20),
    bytes32: new ast_1.FixedBytesType(32),
    stringMemory: new ast_1.PointerType(new ast_1.StringType(), constants_1.DataLocation.Memory),
    bytesMemory: new ast_1.PointerType(new ast_1.BytesType(), constants_1.DataLocation.Memory),
    bytesCalldata: new ast_1.PointerType(new ast_1.BytesType(), constants_1.DataLocation.CallData),
    address: new ast_1.AddressType(false),
    addressPayable: new ast_1.AddressType(true),
    noType: new ast_1.TupleType([]),
    typeOfType: new ast_1.BuiltinFunctionType(undefined, [], [])
};
