// solc-typed-ast/dist/ast/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PossibleTimeUnits = exports.PossibleEtherUnits = exports.PossibleLiteralKinds = exports.PossibleFunctionCallKinds = exports.PossibleFunctionKinds = exports.PossibleContractKinds = exports.PossibleMutabilities = exports.PossibleStateVariableVisibilities = exports.PossibleFunctionStateMutabilities = exports.PossibleFunctionVisibilities = exports.PossibleDataLocations = exports.RawCommentKind = exports.TimeUnit = exports.EtherUnit = exports.LiteralKind = exports.ModifierInvocationKind = exports.FunctionCallKind = exports.FunctionKind = exports.ContractKind = exports.Mutability = exports.StateVariableVisibility = exports.FunctionStateMutability = exports.FunctionVisibility = exports.DataLocation = exports.ExternalReferenceType = void 0;
var ExternalReferenceType;
(function (ExternalReferenceType) {
    ExternalReferenceType["UserDefined"] = "userDefined";
    ExternalReferenceType["Builtin"] = "builtin";
})(ExternalReferenceType || (exports.ExternalReferenceType = ExternalReferenceType = {}));
var DataLocation;
(function (DataLocation) {
    DataLocation["Storage"] = "storage";
    DataLocation["Memory"] = "memory";
    DataLocation["CallData"] = "calldata";
    DataLocation["Default"] = "default";
})(DataLocation || (exports.DataLocation = DataLocation = {}));
var FunctionVisibility;
(function (FunctionVisibility) {
    FunctionVisibility["External"] = "external";
    FunctionVisibility["Public"] = "public";
    FunctionVisibility["Internal"] = "internal";
    FunctionVisibility["Private"] = "private";
    FunctionVisibility["Default"] = "default";
})(FunctionVisibility || (exports.FunctionVisibility = FunctionVisibility = {}));
var FunctionStateMutability;
(function (FunctionStateMutability) {
    FunctionStateMutability["Pure"] = "pure";
    FunctionStateMutability["View"] = "view";
    FunctionStateMutability["Constant"] = "constant";
    FunctionStateMutability["Payable"] = "payable";
    FunctionStateMutability["NonPayable"] = "nonpayable";
})(FunctionStateMutability || (exports.FunctionStateMutability = FunctionStateMutability = {}));
var StateVariableVisibility;
(function (StateVariableVisibility) {
    StateVariableVisibility["Public"] = "public";
    StateVariableVisibility["Internal"] = "internal";
    StateVariableVisibility["Private"] = "private";
    StateVariableVisibility["Default"] = "default";
})(StateVariableVisibility || (exports.StateVariableVisibility = StateVariableVisibility = {}));
var Mutability;
(function (Mutability) {
    Mutability["Mutable"] = "mutable";
    Mutability["Immutable"] = "immutable";
    Mutability["Constant"] = "constant";
})(Mutability || (exports.Mutability = Mutability = {}));
var ContractKind;
(function (ContractKind) {
    ContractKind["Contract"] = "contract";
    ContractKind["Library"] = "library";
    ContractKind["Interface"] = "interface";
})(ContractKind || (exports.ContractKind = ContractKind = {}));
var FunctionKind;
(function (FunctionKind) {
    FunctionKind["Constructor"] = "constructor";
    FunctionKind["Function"] = "function";
    FunctionKind["Receive"] = "receive";
    FunctionKind["Fallback"] = "fallback";
    FunctionKind["Free"] = "freeFunction";
})(FunctionKind || (exports.FunctionKind = FunctionKind = {}));
var FunctionCallKind;
(function (FunctionCallKind) {
    FunctionCallKind["FunctionCall"] = "functionCall";
    FunctionCallKind["TypeConversion"] = "typeConversion";
    FunctionCallKind["StructConstructorCall"] = "structConstructorCall";
})(FunctionCallKind || (exports.FunctionCallKind = FunctionCallKind = {}));
var ModifierInvocationKind;
(function (ModifierInvocationKind) {
    ModifierInvocationKind["ModifierInvocation"] = "modifierInvocation";
    ModifierInvocationKind["BaseConstructorSpecifier"] = "baseConstructorSpecifier";
})(ModifierInvocationKind || (exports.ModifierInvocationKind = ModifierInvocationKind = {}));
var LiteralKind;
(function (LiteralKind) {
    LiteralKind["Number"] = "number";
    LiteralKind["Bool"] = "bool";
    LiteralKind["String"] = "string";
    LiteralKind["HexString"] = "hexString";
    LiteralKind["UnicodeString"] = "unicodeString";
})(LiteralKind || (exports.LiteralKind = LiteralKind = {}));
var EtherUnit;
(function (EtherUnit) {
    EtherUnit["Wei"] = "wei";
    EtherUnit["GWei"] = "gwei";
    EtherUnit["Szabo"] = "szabo";
    EtherUnit["Finney"] = "finney";
    EtherUnit["Ether"] = "ether";
})(EtherUnit || (exports.EtherUnit = EtherUnit = {}));
var TimeUnit;
(function (TimeUnit) {
    TimeUnit["Seconds"] = "seconds";
    TimeUnit["Minutes"] = "minutes";
    TimeUnit["Hours"] = "hours";
    TimeUnit["Days"] = "days";
    TimeUnit["Weeks"] = "weeks";
    TimeUnit["Years"] = "years";
})(TimeUnit || (exports.TimeUnit = TimeUnit = {}));
var RawCommentKind;
(function (RawCommentKind) {
    RawCommentKind["SingleLineComment"] = "single_line";
    RawCommentKind["BlockComment"] = "block_comment";
    RawCommentKind["LineGroupNatSpec"] = "line_group_natspec";
    RawCommentKind["BlockNatSpec"] = "block_natspec";
})(RawCommentKind || (exports.RawCommentKind = RawCommentKind = {}));
exports.PossibleDataLocations = new Set(Object.values(DataLocation));
exports.PossibleFunctionVisibilities = new Set(Object.values(FunctionVisibility));
exports.PossibleFunctionStateMutabilities = new Set(Object.values(FunctionStateMutability));
exports.PossibleStateVariableVisibilities = new Set(Object.values(StateVariableVisibility));
exports.PossibleMutabilities = new Set(Object.values(Mutability));
exports.PossibleContractKinds = new Set(Object.values(ContractKind));
exports.PossibleFunctionKinds = new Set(Object.values(FunctionKind));
exports.PossibleFunctionCallKinds = new Set(Object.values(FunctionCallKind));
exports.PossibleLiteralKinds = new Set(Object.values(LiteralKind));
exports.PossibleEtherUnits = new Set(Object.values(EtherUnit));
exports.PossibleTimeUnits = new Set(Object.values(TimeUnit));
