// solc-typed-ast/dist/compile/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PossibleCompilationOutputs = exports.PossibleCompilerKinds = exports.CompilationOutput = exports.CompilerKind = exports.LatestCompilerVersion = exports.CompilerVersions = exports.CompilerSeries = exports.CompilerVersions08 = exports.CompilerVersions07 = exports.CompilerVersions06 = exports.CompilerVersions05 = exports.CompilerVersions04 = void 0;
exports.CompilerVersions04 = [
    "0.4.13",
    "0.4.14",
    "0.4.15",
    "0.4.16",
    "0.4.17",
    "0.4.18",
    "0.4.19",
    "0.4.20",
    "0.4.21",
    "0.4.22",
    "0.4.23",
    "0.4.24",
    "0.4.25",
    "0.4.26"
];
exports.CompilerVersions05 = [
    "0.5.0",
    "0.5.1",
    "0.5.2",
    "0.5.3",
    "0.5.4",
    "0.5.5",
    "0.5.6",
    "0.5.7",
    "0.5.8",
    "0.5.9",
    "0.5.10",
    "0.5.11",
    "0.5.12",
    "0.5.13",
    "0.5.14",
    "0.5.15",
    "0.5.16",
    "0.5.17"
];
exports.CompilerVersions06 = [
    "0.6.0",
    "0.6.1",
    "0.6.2",
    "0.6.3",
    "0.6.4",
    "0.6.5",
    "0.6.6",
    "0.6.7",
    "0.6.8",
    "0.6.9",
    "0.6.10",
    "0.6.11",
    "0.6.12"
];
exports.CompilerVersions07 = ["0.7.0", "0.7.1", "0.7.2", "0.7.3", "0.7.4", "0.7.5", "0.7.6"];
exports.CompilerVersions08 = [
    "0.8.0",
    "0.8.1",
    "0.8.2",
    "0.8.3",
    "0.8.4",
    "0.8.5",
    "0.8.6",
    "0.8.7",
    "0.8.8",
    "0.8.9",
    "0.8.10",
    "0.8.11",
    "0.8.12",
    "0.8.13",
    "0.8.14",
    "0.8.15",
    "0.8.16",
    "0.8.17",
    "0.8.18",
    "0.8.19",
    "0.8.20",
    "0.8.21",
    "0.8.22",
    "0.8.23"
];
exports.CompilerSeries = [
    exports.CompilerVersions04,
    exports.CompilerVersions05,
    exports.CompilerVersions06,
    exports.CompilerVersions07,
    exports.CompilerVersions08
];
exports.CompilerVersions = [
    ...exports.CompilerVersions04,
    ...exports.CompilerVersions05,
    ...exports.CompilerVersions06,
    ...exports.CompilerVersions07,
    ...exports.CompilerVersions08
];
exports.LatestCompilerVersion = exports.CompilerVersions[exports.CompilerVersions.length - 1];
var CompilerKind;
(function (CompilerKind) {
    CompilerKind["WASM"] = "wasm";
    CompilerKind["Native"] = "native";
})(CompilerKind || (exports.CompilerKind = CompilerKind = {}));
/**
 * Corresponds to the string constants used in "outputSelection" as described in
 * https://docs.soliditylang.org/en/latest/using-the-compiler.html#input-description
 */
var CompilationOutput;
(function (CompilationOutput) {
    CompilationOutput["AST"] = "ast";
    CompilationOutput["ABI"] = "abi";
    CompilationOutput["DEVDOC"] = "devdoc";
    CompilationOutput["USERDOC"] = "userdoc";
    CompilationOutput["METADATA"] = "metadata";
    CompilationOutput["IR"] = "ir";
    CompilationOutput["IR_OPTIMIZED"] = "irOptimized";
    CompilationOutput["STORAGE_LAYOUT"] = "storageLayout";
    CompilationOutput["EVM"] = "evm";
    CompilationOutput["EVM_ASSEMBLY"] = "evm.assembly";
    CompilationOutput["EVM_LEGACY_ASSEMBLY"] = "evm.legacyAssembly";
    CompilationOutput["EVM_BYTECODE"] = "evm.bytecode";
    CompilationOutput["EVM_BYTECODE_OBJECT"] = "evm.bytecode.object";
    CompilationOutput["EVM_BYTECODE_OPCODES"] = "evm.bytecode.opcodes";
    CompilationOutput["EVM_BYTECODE_SOURCEMAP"] = "evm.bytecode.sourceMap";
    CompilationOutput["EVM_BYTECODE_LINKREFERENCES"] = "evm.bytecode.linkReferences";
    CompilationOutput["EVM_BYTECODE_GENERATEDSOURCES"] = "evm.bytecode.generatedSources";
    CompilationOutput["EVM_DEPLOYEDBYTECODE_IMMUTABLEREFERENCES"] = "evm.deployedBytecode.immutableReferences";
    CompilationOutput["EVM_METHODIDENTIFIERS"] = "evm.methodIdentifiers";
    CompilationOutput["EVM_GASESTIMATES"] = "evm.gasEstimates";
    CompilationOutput["EWASM_WAST"] = "ewasm.wast";
    CompilationOutput["EWASM_WASM"] = "ewasm.wasm";
    CompilationOutput["ALL"] = "*";
})(CompilationOutput || (exports.CompilationOutput = CompilationOutput = {}));
exports.PossibleCompilerKinds = new Set(Object.values(CompilerKind));
exports.PossibleCompilationOutputs = new Set(Object.values(CompilationOutput));
