// solc-typed-ast/dist/ast/implementation/meta/using_for_directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsingForDirective = void 0;
const ast_node_1 = require("../../ast_node");
const identifier_path_1 = require("./identifier_path");
class UsingForDirective extends ast_node_1.ASTNode {
    constructor(id, src, isGlobal, libraryName, functionList, typeName, raw) {
        super(id, src, raw);
        if (libraryName) {
            this.vLibraryName = libraryName;
        }
        else if (functionList) {
            this.vFunctionList = functionList;
        }
        else {
            throw new Error("One of vLibraryName or vFunctionList should always be set");
        }
        this.vTypeName = typeName;
        this.isGlobal = isGlobal;
        this.acceptChildren();
    }
    get children() {
        const result = [];
        if (this.vLibraryName) {
            result.push(this.vLibraryName);
        }
        if (this.vFunctionList) {
            for (const entry of this.vFunctionList) {
                result.push(entry instanceof identifier_path_1.IdentifierPath ? entry : entry.definition);
            }
        }
        if (this.vTypeName) {
            result.push(this.vTypeName);
        }
        return result;
    }
}
exports.UsingForDirective = UsingForDirective;
