// solc-typed-ast/dist/ast/implementation/type/user_defined_type_name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserDefinedTypeName = void 0;
const type_name_1 = require("./type_name");
class UserDefinedTypeName extends type_name_1.TypeName {
    constructor(id, src, typeString, name, referencedDeclaration, path, raw) {
        super(id, src, typeString, raw);
        this.name = name;
        this.referencedDeclaration = referencedDeclaration;
        this.path = path;
        this.acceptChildren();
    }
    get children() {
        return this.pickNodes(this.path);
    }
    /**
     * Reference to the declaration
     */
    get vReferencedDeclaration() {
        return this.requiredContext.locate(this.referencedDeclaration);
    }
    set vReferencedDeclaration(value) {
        if (!this.requiredContext.contains(value)) {
            throw new Error(`Node ${value.type}#${value.id} not belongs to a current context`);
        }
        this.referencedDeclaration = value.id;
    }
}
exports.UserDefinedTypeName = UserDefinedTypeName;
